cupy/cuda/cupy_thrust.cu

```diff
@@ -17,29 +18,31 @@ typedef hipStream_t cudaStream_t;
 namespace cuda {
     using thrust::hip::par;
 }
-
 #endif // #if CUPY_USE_HIP
 
 
-extern "C" char *cupy_malloc(void *, ptrdiff_t);
-extern "C" void cupy_free(void *, char *);
-
-
 class cupy_allocator {
 private:
-    void* memory;
+    cupy_allocator_handle* pool = nullptr;
 
 public:
     typedef char value_type;
 
-    cupy_allocator(void* memory) : memory(memory) {}
+    cupy_allocator() {
+        pool = get_cupy_allocator_handle();
+    }
+
+    ~cupy_allocator() {
+        destroy_cupy_allocator_handle(pool);
+    }
 
-    char *allocate(std::ptrdiff_t num_bytes) {
-        return cupy_malloc(memory, num_bytes);
+    char* allocate(size_t num_bytes) {
+        return (char*)cupy_malloc(pool, num_bytes);
```

Comment:
I think that calling directly `cupy_malloc` is not safe.
Probably we should add a new `cupy_malloc(pool, size, stream)` to allocate memory on a given stream, this should make it versatile.

Reply:
Hi Emilio, "not safe" because of the lack of a stream? I thought that none of our allocation routines accept a stream pointer, and that it's acquired internally by looking up the thread-local storage?
https://github.com/cupy/cupy/blob/ff18a56f23058a947e5ceab74995deda8514b2f1/cupy/cuda/memory.pyx#L940
Note that the original code has no stream argument either.

In any case, it's not possible to declare `cupy_malloc(pool, size, stream)` and then wrap it with `allocate(size, stream)` on top, because the signature of `allocate` is required by Thrust to be one or two arguments. In the latter case, it's for alignment AFAIK. If we need the current stream, we should look it up in the Cython space, not in here. (And this circles back to my point above.)

Reply:
The stream that your external library is using might be different from the current cupy one.

Reply:
Ah, I see what you're saying. OK, for `cupy/cuda/cupy_thrust.cu` there's nothing to fix, because we pass in the current stream from CuPy. So your point is on the general applicability of the new C API. Let me think about it. Thanks for bringing this up, Emilio!

Reply:
Let me know if I miss something here. There're 4 possibilities during allocation:
1. External library is not using any stream: The arena would be recorded on the null stream.
2. External library is using its own stream (perhaps in the C/C++ space): Same as in 1, the arena is recorded on the null stream, since CuPy has no way to know the external stream.
3. External library is using a CuPy stream: In this case, the CuPy stream can be picked up internally.
4. External library is using its own stream (either in C/C++ or Python) *and* a CuPy stream in Python: The arena is recorded on the CuPy stream. 

Synchronization issues aside, it seems to me if during deallocation the stream usage remains unchanged (i.e., using the same stream as in allocation), we will be able to locate the correct arena?